rewrite-maven/src/main/java/org/openrewrite/maven/UpgradeDependencyVersion.java

```diff
@@ -125,35 +126,81 @@ public String getDescription() {
     }
 
     @Override
-    public Set<GroupArtifact> getInitialValue(ExecutionContext ctx) {
-        return new HashSet<>();
+    public Accumulator getInitialValue(ExecutionContext ctx) {
+        return new Accumulator();
     }
 
     @Override
-    public TreeVisitor<?, ExecutionContext> getScanner(Set<GroupArtifact> projectArtifacts) {
+    public TreeVisitor<?, ExecutionContext> getScanner(Accumulator accumulator) {
         return new MavenIsoVisitor<ExecutionContext>() {
+            private final VersionComparator versionComparator =
+                    requireNonNull(Semver.validate(newVersion, versionPattern).getValue());
+
             @Override
             public Xml.Document visitDocument(Xml.Document document, ExecutionContext ctx) {
                 ResolvedPom pom = getResolutionResult().getPom();
-                projectArtifacts.add(new GroupArtifact(pom.getGroupId(), pom.getArtifactId()));
-                return document;
+                accumulator.projectArtifacts.add(new GroupArtifact(pom.getGroupId(), pom.getArtifactId()));
+                return super.visitDocument(document, ctx);
+            }
+
+            @Override
+            public Xml.Tag visitTag(final Xml.Tag tag, final ExecutionContext ctx) {
+                ResolvedDependency d = findDependency(tag);
+                if (d != null && d.getRepository() != null) {
+                    // if the resolved dependency exists AND it does not represent an artifact that was parsed
+                    // as a source file, attempt to find a new version.
+                    try {
+                        String newerVersion = findNewerVersion(d.getVersion(),
+                                () -> downloadMetadata(d.getGroupId(), d.getArtifactId(), ctx), versionComparator, ctx);
```

Comment:
Something I don't quite understand from looking over the code is that we appear to be looking at _any _dependency, not just dependencies whose groupId & artifactId match the one we're looking to upgrade. Am I reading that correctly?